src/duckdb.mjs

```diff
@@ -1,361 +1,228 @@
+// POSSIBILITY OF SUCH DAMAGE.
+
+// TODO Allow this to be overridden using the Library’s resolver.
+const cdn = "https://cdn.observableusercontent.com/npm/";
+
+export class DuckDBClient {
+  constructor(db) {
+    Object.defineProperties(this, {
+      _db: {value: db}
+    });
   }
 
-  // Adapted from: https://observablehq.com/@cmudig/duckdb-client
-  // Follows the DatabaseClient specification: https://observablehq.com/@observablehq/database-client-specification
-  class DuckDBClient {
-    constructor(_db) {
-      this._db = _db;
-      this._counter = 0;
+  async queryStream(query, params) {
+    const connection = await this._db.connect();
+    let reader, schema, batch;
+    try {
+      reader = await connection.send(query, params);
+      batch = await reader.next();
+      if (batch.done) throw new Error("missing first batch");
+      schema = batch.value.schema;
+    } catch (error) {
+      await connection.close();
+      throw error;
     }
-
-    async queryStream(query, params) {
-      const conn = await this.connection();
-      let result;
-
-      if (params) {
-        const stmt = await conn.prepare(query);
-        result = await stmt.query(...params);
-      } else {
-        result = await conn.query(query);
-      }
-      // Populate the schema of the results
-      const schema = result.schema.fields.map(({
-        name,
-        type
-      }) => ({
+    return {
+      schema: schema.fields.map(({name, type}) => ({
         name,
         type: getType(String(type)),
         databaseType: String(type)
-      }));
-      return {
-        schema,
-        async *readRows() {
-          let rows = result.toArray().map((r) => Object.fromEntries(r));
-          yield rows;
-        }
-      };
-    }
-
-    // This function gets called to prepare the `query` parameter of the `queryStream` method
-    queryTag(strings, ...params) {
-      return [strings.join("?"), params];
-    }
-
-    escape(name) {
-      return `"${name}"`;
-    }
-
-    async describeTables() {
-      const conn = await this.connection();
-      const tables = (await conn.query(`SHOW TABLES`)).toArray();
-      return tables.map(({
-        name
-      }) => ({
-        name
-      }));
-    }
-
-    async describeColumns({
-      table
-    } = {}) {
-      const conn = await this.connection();
-      const columns = (await conn.query(`DESCRIBE ${table}`)).toArray();
-      return columns.map(({
-        column_name,
-        column_type
-      }) => {
-        return {
-          name: column_name,
-          type: getType(column_type),
-          databaseType: column_type
-        };
-      });
-    }
-
-    async db() {
-      if (!this._db) {
-        this._db = await makeDB();
-        await this._db.open({
-          query: {
-            castTimestampToDate: true
+      })),
+      async *readRows() {
+        try {
+          while (!batch.done) {
+            yield batch.value.toArray();
```

Comment:
Not as part of this PR, but we should have an option to return Apache Arrow Table instances directly from the database client, rather than converting to an array of objects (rows) here. Probably this is an option (arrow = true) when you run a query, and perhaps you can specify it as a default option when constructing the DuckDBClient. In conjunction with #315 it would mean that we could entirely avoid materializing the array of objects when displaying the results of a DuckDBClient query in a table.

Reply:
At the moment `DuckDBClient.query` returns Apache Arrow Table so this sounds good once again.

Reply:
And It would be really great if you could use Apache Arrow Table also with Plot.

Reply:
@kimmolinna Agreed! Tracking that here: https://github.com/observablehq/plot/issues/1103

Reply:
Sweet. I think native arrow support throughout would be awesome since we reduce materializations.